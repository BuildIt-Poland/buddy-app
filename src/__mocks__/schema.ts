export enum UserRole {
  Newbie = 'NEWBIE',
  Buddy = 'BUDDY',
}

export enum TaskStatus {
  Completed = 'COMPLETED',
  Uncompleted = 'UNCOMPLETED',
}

export default {
  UserRole,
  TaskStatus,
};
