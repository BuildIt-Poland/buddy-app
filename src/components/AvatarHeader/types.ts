export interface AvatarHeaderProps {
  newbieId: string;
}
