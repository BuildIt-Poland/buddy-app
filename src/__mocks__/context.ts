import { AuthContextData } from 'context/AuthStore';

export const authContext = (context?: Partial<AuthContextData>) => ({
  data: { role: 'BUDDY', token: 'dummy-token', userId: '1' },
  loading: false,
  error: null,
  isAuthenticated: true,
  login: jest.fn(),
  logout: jest.fn(),
  ...context,
});
