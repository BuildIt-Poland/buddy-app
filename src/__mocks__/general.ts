import { ROUTES } from 'shared/routes';

export const mockLocation = (path: string = ROUTES.BASE) => ({
  key: 'utwyk7',
  pathname: path,
});

export const mockSchemaTypes = (types: any = {}) => ({
  UserRole: {
    Newbie: 'NEWBIE',
    Buddy: 'BUDDY',
  },
  TaskStatus: {
    Completed: 'COMPLETED',
    Uncompleted: 'UNCOMPLETED',
  },
  ...types,
});
