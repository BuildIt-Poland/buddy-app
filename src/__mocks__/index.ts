export * from './general';
export * from './graphql';
export * from './schema';
