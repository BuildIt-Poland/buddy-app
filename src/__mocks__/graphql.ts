import { LOGIN_MUTATION } from 'graphql/login.graphql';
import { BUDDY_MENU_DETAILS, NEWBIE_MENU_DETAILS } from 'graphql/user-menu.graphql';
import { AVATAR_HEADER } from 'graphql/avatar-header.graphql';
import { NEWBIE_SELECT } from 'graphql/newbie-select.graphql';
import { TASK_DETAILS } from 'graphql/task-details.graphql';
import { GraphQLError } from 'graphql';
import {
  MutationLoginArgs,
  AuthPayload,
  TaskStatus,
  Newbie,
  QueryNewbieArgs,
  QueryBuddyArgs,
  UserRole,
} from 'buddy-app-schema';
import {
  NEWBIE_CONTACT_DETAILS,
  BUDDY_CONTACT_DETAILS,
} from 'graphql/contact-details.graphql';
import { getBasicUserDetailsMock } from './general';
import { authContext } from './context';

export const loginSuccessMock = (
  variables?: Partial<MutationLoginArgs>,
  data?: Partial<AuthPayload>
) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  result: {
    data: {
      login: {
        ...authContext().data,
        ...data,
      },
    },
  },
});

export const loginFailedMock = (variables?: Partial<MutationLoginArgs>) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  result: {
    errors: [new GraphQLError('No such user found')],
  },
});

export const loginNoNetworkMock = (variables?: Partial<MutationLoginArgs>) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  error: new Error('Network error'),
});

export const NewbieAvatarDetails = (variables?: QueryNewbieArgs) => ({
  request: {
    query: AVATAR_HEADER,
    variables,
  },
  result: {
    data: {
      newbie: {
        name: 'Tom Hanks',
        position: 'front-end',
        photo: 'some-url',
        tasksInfo: {
          buddyProgress: 0.54,
        },
      },
    },
  },
});

export const tasksResponse = (filter?: TaskStatus) => {
  const tasks = [
    {
      id: '1',
      title: 'Call you head manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Uncompleted,
      newbie: {} as Newbie,
    },
    {
      id: '2',
      title: 'Write you main manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Completed,
      newbie: {} as Newbie,
    },
  ];

  switch (filter) {
    case TaskStatus.Completed:
      return [tasks[0]];
    case TaskStatus.Uncompleted:
      return [tasks[1]];
    default:
      return tasks;
  }
};

export const buddyMenuDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: {
          name: 'Tom Hanks',
          email: 'tom@wipro.com',
          allowPushedNotifications: true,
          photo: 'buddy-photo-url',
          newbies: [
            {
              id: 'id',
              name: 'Sandra Bullock',
              photo: 'newbie-photo-url',
            },
          ],
        },
      },
    },
  },
];

export const newbieMenuDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: {
          name: 'Sandra Bullock',
          email: 'doejoe@wipro.com',
          allowPushedNotifications: true,
          photo: 'newbie-photo-url',
          buddy: {
            id: 'id',
            name: 'Tom Hanks',
            photo: 'newbie-photo-url',
          },
        },
      },
    },
  },
];

export const newbieContactDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: getBasicUserDetailsMock(UserRole.Newbie),
      },
    },
  },
];

export const buddyContactDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: getBasicUserDetailsMock(UserRole.Buddy),
      },
    },
  },
];

export const newbieSelectMock = [
  {
    request: {
      query: NEWBIE_SELECT,
      variables: {
        buddyId: 'ck17sl83c9gya0b17dcvttzm4',
      },
    },
    result: {
      data: {
        buddy: {
          id: 'ck17sl83c9gya0b17dcvttzm4',
          name: 'Dummy',
          role: 'BUDDY',
          photo: null,
          newbies: [
            {
              id: 'ck17svulh9k2k0b17j31ansfk',
              photo: null,
              name: 'Test 2',
              startDate: null,
              tasksInfo: { buddyProgress: 0.5 },
            },
            {
              id: 'ck17swp2m9kcv0b17we0ibrdn',
              photo: null,
              name: 'Test 1',
              startDate: null,
              tasksInfo: { buddyProgress: 0.6 },
            },
          ],
        },
      },
    },
  },
];

export const taskDetailsMock = [
  {
    request: {
      query: TASK_DETAILS,
      variables: {
        taskId: '1',
      },
    },
    result: {
      data: {
        task: {
          title: 'New task',
          description: 'New task',
          status: TaskStatus.Uncompleted,
        },
      },
    },
  },
];
