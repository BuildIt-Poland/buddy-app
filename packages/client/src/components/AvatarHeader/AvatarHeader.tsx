import React from 'react';
import { useHistory } from 'react-router';
import makeStyles from '@material-ui/core/styles/makeStyles';
import CircularProgress from '@material-ui/core/CircularProgress';
import { Theme } from '@material-ui/core/';
import Box from '@material-ui/core/Box';
import { useQuery } from '@apollo/react-hooks';
import Avatar from 'atoms/Avatar';
import { useAuth } from 'contexts/AuthContext';
import { AVATAR_HEADER } from 'graphql/avatar-header.graphql';
import { Query, QueryNewbieArgs, UserRole } from '@buddy-app/schema';
import { getProgressInPercentages } from 'utils';
import { ROUTES } from 'shared/routes';
import { AvatarHeaderProps } from './types';

const useStyles = makeStyles<Theme>(theme => ({
  background: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    minHeight: theme.spacing(20),
  },
}));

const AvatarHeader: React.FC<AvatarHeaderProps> = ({ newbieId, taskProgress }) => {
  const {
    data: { role },
  } = useAuth();
  const history = useHistory();
  const { background } = useStyles();

  const routes = {
    [UserRole.Buddy]: ROUTES.BUDDY_NEWBIE_DETAILS.replace(':newbieId', newbieId),
    [UserRole.Newbie]: ROUTES.NEWBIE_DETAILS,
  };

  const { loading, data } = useQuery<Query, QueryNewbieArgs>(AVATAR_HEADER, {
    variables: { newbieId },
  });

  const handleClick = () => history.push(routes[role]);

  return (
    <Box className={background} data-testid='avatar-header'>
      {data && !loading && (
        <Avatar
          name={data.newbie.name}
          imgSrc={data.newbie.photo}
          position={data.newbie.position || ''}
          progress={getProgressInPercentages(taskProgress)}
          onClick={handleClick}
        />
      )}
      {loading && <CircularProgress />}
    </Box>
  );
};

export default AvatarHeader;
