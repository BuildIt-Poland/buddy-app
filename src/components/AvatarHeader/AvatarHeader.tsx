import React from 'react';
import { useHistory } from 'react-router';
import { Box, createStyles, Theme } from '@material-ui/core';
import makeStyles from '@material-ui/core/styles/makeStyles';
import CircularProgress from '@material-ui/core/CircularProgress';
import { useQuery } from '@apollo/react-hooks';
import Avatar from 'components/Avatar';
import { AVATAR_HEADER } from 'graphql/avatar-header.graphql';

import { Query, QueryNewbieArgs } from 'buddy-app-schema';
import { getProgressInPercentages } from 'utils';
import { ROUTES } from 'shared/routes';
import { AvatarHeaderProps } from './types';

const useStyles = makeStyles<Theme>(theme =>
  createStyles({
    background: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      width: '100%',
      minHeight: theme.spacing(20),
    },
  })
);

const AvatarHeader: React.FC<AvatarHeaderProps> = ({ newbieId }) => {
  const history = useHistory();
  const { background } = useStyles();

  const { loading, data } = useQuery<Query, QueryNewbieArgs>(AVATAR_HEADER, {
    variables: { newbieId },
  });

  const handleClick = () => {
    history.push(ROUTES.BUDDY_NEWBIE_DETAILS.replace(':newbieId', newbieId));
  };

  return (
    <Box className={background} data-testid='avatar-header'>
      {data && !loading && (
        <Avatar
          name={data.newbie.name}
          imgSrc={data.newbie.photo}
          position={data.newbie.position || ''}
          progress={getProgressInPercentages(data.newbie.tasksInfo.buddyProgress)}
          onClick={handleClick}
        />
      )}
      {loading && <CircularProgress />}
    </Box>
  );
};

export default AvatarHeader;
