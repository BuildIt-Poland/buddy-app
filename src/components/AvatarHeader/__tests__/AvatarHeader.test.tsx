import React from 'react';
import { act, create, ReactTestRenderer } from 'react-test-renderer';

import AvatarHeader from 'components/AvatarHeader';
import { MockedProvider } from '@apollo/react-testing';
import { MemoryRouter, Route } from 'react-router';
import waitForExpect from 'wait-for-expect';
import { NewbieAvatarDetails } from '__mocks__';

jest.mock('@material-ui/core/Box', () => 'Box');
jest.mock('@material-ui/core/CircularProgress', () => 'CircularProgress');
jest.mock('components/Avatar', () => 'Avatar');
jest.mock('buddy-app-schema', () => {});

describe('Component - AvatarHeader', () => {
  const path = '/buddy/newbies/1234/tasks';

  it('renders correctly', async () => {
    let component: ReactTestRenderer;
    component = create(
      <MockedProvider
        mocks={[NewbieAvatarDetails({ newbieId: '1234' })]}
        addTypename={false}>
        <MemoryRouter initialEntries={[path]}>
          <Route path={'/buddy/newbies/:newbieId/tasks'}>
            <AvatarHeader />
          </Route>
        </MemoryRouter>
      </MockedProvider>
    );
    expect(component).toMatchSnapshot();

    await act(async () => {
      await waitForExpect(() => {
        expect(component).toMatchSnapshot();
      });
    });
  });
});
