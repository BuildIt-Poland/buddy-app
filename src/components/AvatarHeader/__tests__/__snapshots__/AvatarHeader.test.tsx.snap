// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - AvatarHeader renders correctly 1`] = `
<Box
  className="makeStyles-background-1"
  data-testid="avatar-header"
>
  <CircularProgress />
</Box>
`;

exports[`Component - AvatarHeader renders correctly 2`] = `
<Box
  className="makeStyles-background-1"
  data-testid="avatar-header"
/>
`;
